src/audio_buffer.rs

```diff
@@ -0,0 +1,482 @@
+        let mut channels = Vec::<Arc<Vec<f32>>>::with_capacity(number_of_channels);
+        // [spec] Note: This initializes the underlying storage to zero.
+        // we don't want to clone just the Arc here, but the whole underlying data,
+        // so we use `fill_with` instead of `fill`
+        // internal_data.fill_with(|| Arc::new(vec![0.; options.length as usize]));
+        for _ in 0..number_of_channels {
+            let mut channel = Vec::<f32>::with_capacity(length);
+            channel.resize(10, 0.);
+
+            channels.push(Arc::new(channel));
+        }
+
+        Self {
+            channels: Rc::new(channels),
+        }
+    }
+}
+
+// used in `decodeAudioData`
+impl From<Vec<Vec<f32>>> for AudioBufferData {
+    fn from(decoded: Vec<Vec<f32>>) -> Self {
+        let number_of_channels = decoded.len();
+        let mut channels = Vec::<Arc<Vec<f32>>>::with_capacity(number_of_channels);
+
+        for channel in decoded.iter() {
+            // @note - is there a better way than copy using `to_vec`?
+            // basically, is there any way to grab the `decoded.channel` reference and
+            // put it where we want it to be without memory allocation?
+            // (sometimes JS feels so simple... :)
+            let copy = channel.to_vec();
+            channels.push(Arc::new(copy));
+        }
+
+        Self {
+            channels: Rc::new(channels),
+        }
+    }
+}
+
+// @note - what could be default/required values in Rust syntax, is this possible?
+#[derive(Copy, Clone, Debug)]
+pub struct AudioBufferOptions {
+    number_of_channels: usize, // defaults to 1
+    length: usize,             // required
+    sample_rate: SampleRate,   // required
+}
+
+#[derive(Clone)]
+pub struct AudioBuffer {
+    number_of_channels: usize,
+    length: usize,
+    sample_rate: SampleRate,
+    internal_data: AudioBufferData,
+}
+
+// @todo - possible solution to make the `offlineContext.startRendering` API compliant
+// impl From<buffer::AudioBuffer> for AudioBuffer {}
+
+impl AudioBuffer {
+    // https://webaudio.github.io/web-audio-api/#AudioBuffer-constructors
+    pub fn new(options: AudioBufferOptions) -> Self {
+        let AudioBufferOptions {
+            number_of_channels,
+            length,
+            sample_rate,
+        } = options;
+
+        let internal_data = AudioBufferData::new(number_of_channels, length);
+
+        Self {
+            number_of_channels,
+            length,
+            sample_rate,
+            internal_data,
+        }
+    }
+
+    pub fn number_of_channels(&self) -> usize {
+        self.number_of_channels
+    }
+
+    pub fn length(&self) -> usize {
+        self.length
+    }
+
+    pub fn sample_rate(&self) -> SampleRate {
+        self.sample_rate
+    }
+
+    pub fn duration(&self) -> f64 {
+        (self.length / self.sample_rate.0 as usize) as f64
+    }
+
+    // @note - not sure if we can handle default arguments in another way
```

Comment:
This is fine, or take an Optional argument

Reply:
No real opinion on that (I don't really like having None everywhere but...), probably the important thing is to be consistent within the library